### PR TITLE
ci: add wheel artifact parity check against source package

### LINKED ISSUE #2354: CI: add wheel artifact parity check against the source package
## Summary

The installed `arnio==1.19.0` wheel is out of sync with the current source package. A release/artifact parity check should prevent wheels from missing Python modules or shipping stale module contents.

## Evidence

Comparing the repository package at:

```text
C:\Users\anish\Desktop\arnio\arnio
```

with the installed wheel package at:

```text
C:\Users\anish\Desktop\arnio\.audit_venv\Lib\site-packages\arnio
```

showed:

```json
{
  "repo_py_count": 18,
  "inst_py_count": 15,
  "missing_in_wheel": [
    "_version.py",
    "cli.py",
    "integrations/polars.py"
  ],
  "changed_common_count": 14,
  "changed_common_sample": [
    "__init__.py",
    "cleaning.py",
    "convert.py",
    "exceptions.py",
    "frame.py",
    "integrations/__init__.py",
    "integrations/duckdb.py",
    "integrations/pandas.py",
    "integrations/sklearn.py",
    "io.py",
    "pipeline.py",
    "quality.py",
    "schema.py",
    "schema_export.py"
  ]
}
```

This explains multiple user-facing package regressions, including missing CLI entry points and missing optional integration exports.

## Impact

A wheel can pass basic import checks while still missing important Python modules, examples, or integration exports. Users installing from PyPI then see behavior that differs from the repository docs and tests.

## Suggested Fix

Add a release/wheel smoke step that builds or installs the wheel and verifies package-file parity for expected Python modules. The check should at minimum confirm that all runtime `.py` files under `arnio/` that are meant to ship are present in the installed wheel.

## Acceptance Criteria

- CI or release workflow builds/installs the wheel in a clean environment.
- The check verifies expected runtime modules are present, including `arnio.cli`, `arnio._version`, and optional integration wrappers such as `arnio.integrations.polars`.
- The check verifies important exports from `arnio.__all__` and `arnio.integrations.__all__` match source expectations.
- The check fails if source package files are missing from the built wheel.
- Document any intentionally excluded files so the parity check has an explicit allowlist.

## Related Findings

- #2351
- #2352
- #2353

### PR DESCRIPTION
**Pull request by AaryanInzalkar**
## Summary

Fixes #2354

The published arnio wheel was missing `_version.py`, `cli.py`, and `integrations/polars.py` compared to the source tree, causing user-facing regressions (#2351, #2352, #2353). This PR adds a CI gate that prevents that from happening again.

## Changes

- `scripts/check_wheel_parity.py` — a standalone script that:
  - Builds the wheel from source and installs it into a clean venv
  - Checks that every runtime `.py` file under `arnio/` is present in the installed wheel
  - Verifies all required modules are importable: `arnio.cli`, `arnio._version`, `arnio.integrations.polars`, and others
  - Verifies `arnio.__all__` and `arnio.integrations.__all__` contain expected public symbols
  - Includes an explicit `EXCLUDE_ALLOWLIST` for intentionally excluded files
  - Exits with code 1 on any failure so CI blocks the merge

- `.github/workflows/wheel-parity.yml` — GitHub Actions workflow that:
  - Runs on push/PR to `main` and `release/**` when source or packaging files change
  - Runs across Ubuntu, Windows, and macOS on Python 3.9, 3.11, and 3.13
  - Supports `workflow_call` so the release workflow can use it as a publish gate
  - Uploads the built wheel as a CI artifact for inspection

## Acceptance Criteria Checklist

- [x] CI builds/installs the wheel in a clean environment
- [x] Check verifies `arnio.cli`, `arnio._version`, and `arnio.integrations.polars` are present
- [x] Check verifies `arnio.__all__` and `arnio.integrations.__all__` match source expectations
- [x] Check fails if source `.py` files are missing from the built wheel
- [x] Intentionally excluded files are documented via `EXCLUDE_ALLOWLIST` in the script

### CODE REVIEW

**Comment by im-anishraj** [comment]
@AaryanInzalkar thanks for opening this against #2354. I added the GSSoC tracking labels and marked the PR `status:blocked` for now because the checks are not review-ready yet.

Current blockers:

- `Guard Sensitive Paths and Root Files` is failing because this PR adds `.github/workflows/wheel-parity.yml`. That is expected for workflow changes and requires explicit maintainer review before merge.
- `Lint (Black, Ruff, clang-format)` is failing.
- The new wheel parity matrix is failing on multiple Python/platform jobs.

Please fix the lint/parity failures first, then leave a note with the relevant failure summary and what changed. Keep the workflow diff as small as possible; once the normal checks are clean, a maintainer can decide how to handle the workflow-file guard.

**Comment by im-anishraj** [comment]
@AaryanInzalkar thanks for the quick update. I rechecked the latest commit and the PR guard is now green, but the new wheel-parity job is still failing across the matrix.

The main issue appears to be in `scripts/check_wheel_parity.py`: after installing the wheel into a fresh venv, the subprocess import checks still run from the repository root, so Python can import the source-tree `arnio/` package instead of the wheel-installed package. That source tree does not contain the built `_arnio_cpp` extension, which is why the parity logs show `arnio C++ extension (_arnio_cpp) not found` even though the wheel was built.

Please update the parity script so all installed-wheel inspections run from outside the repository root, or otherwise remove the repo root/current directory from `sys.path` before importing. After that, rerun the matrix and include the new failure summary if anything remains. I am keeping this blocked until the wheel-parity failures are resolved and the normal CI run completes cleanly.

**Comment by AaryanInzalkar** [comment]
> @AaryanInzalkar thanks for the quick update. I rechecked the latest commit and the PR guard is now green, but the new wheel-parity job is still failing across the matrix.
> 
> The main issue appears to be in `scripts/check_wheel_parity.py`: after installing the wheel into a fresh venv, the subprocess import checks still run from the repository root, so Python can import the source-tree `arnio/` package instead of the wheel-installed package. That source tree does not contain the built `_arnio_cpp` extension, which is why the parity logs show `arnio C++ extension (_arnio_cpp) not found` even though the wheel was built.
> 
> Please update the parity script so all installed-wheel inspections run from outside the repository root, or otherwise remove the repo root/current directory from `sys.path` before importing. After that, rerun the matrix and include the new failure summary if anything remains. I am keeping this blocked until the wheel-parity failures are resolved and the normal CI run completes cleanly.

got it i will work on that

**Comment by im-anishraj** [comment]
@AaryanInzalkar thanks for the update. I rechecked the latest commit: normal CI is green now and the parity failures are narrowed to Windows, which is good progress.

The remaining Windows failures are still actionable. In the `Run wheel parity check` step, Windows is passing the glob literally, so the script receives `dist/*.whl` and prints `ERROR: wheel not found: dist\*.whl`. Please make this cross-platform before the check runs, for example by resolving the wheel path inside `scripts/check_wheel_parity.py` when the argument contains a wildcard, or by changing the workflow step to pass an explicit wheel filename on Windows.

Keeping this `status:blocked` until the Windows parity matrix is green. The Vercel authorization failure is external noise and is not the blocker here.

**Comment by AaryanInzalkar** [comment]
@im-anishraj Heyy pls check the CI is green now, Let me know if it needs anymore changes

**Review by im-anishraj — CHANGES_REQUESTED** [review]
@AaryanInzalkar, thank you for resolving the source-tree shadowing and Windows wheel-path problems. I reviewed the full current diff and history, confirmed the entire parity matrix and normal non-Vercel CI are green, and locally verified that the Python script compiles and the workflow YAML parses. One merge blocker remains: the new workflow uses mutable action tags (`actions/checkout@v4`, `setup-python@v5`, `upload-artifact@v4`, and `github-script@v7`). Arnio pins third-party actions to immutable commit SHAs throughout `.github/workflows`; please use the repository current pinned SHAs (with version comments) here as well, and add the missing final newline to `wheel-parity.yml`. Once that is updated and CI remains green, this should be ready for final approval.

**Review by im-anishraj — APPROVED** [review]
@AaryanInzalkar the latest head pins every third-party action to an immutable SHA, restores the final newline, and the full source/wheel parity matrix plus normal non-Vercel CI is green across Linux, Windows, and macOS. Approved.
